### PR TITLE
Add option to resolve polytomies only when the way to introduce splits is unique

### PR DESCRIPTION
**Pull request by PierreBarrat**

### CODE REVIEW

File: src/mcc_branch_labels.jl [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    end
+    return mcc_map
+end
+
+"""
+PRT!(n::TreeNode)
+
+Assign `mcc`s to branches (i.e. their child node) by a Pre-order traversal starting at the root node `n`.
+"""
+function PRT!(n::TreeNode, k::Int)
```
Comment (PierreBarrat) [inline src/mcc_branch_labels.jl]: `PRT!` is not a great name for a function that assigns MCCs to branches ... `assign_mccs_to_branches!`? 
Also, what's `k`?

File: src/mcc_branch_labels.jl [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    end
+
+end
+
+function PRT!(t::Tree)
+	PRT!(t.root)
+end
+
+"""
+function assign_mccs!(mcc_map::Dict{String, Int}, t::Vector{Tree{TreeTools.MiscData}})
```
Comment (PierreBarrat) [inline src/mcc_branch_labels.jl]: As far as I understand, this is modifying the tree `t`, so it should be `assign_mccs!(t, mcc_map)`

File: src/mcc_branch_labels.jl [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+PRT!(n::TreeNode)
+
+Assign `mcc`s to branches (i.e. their child node) by a Pre-order traversal starting at the root node `n`.
+"""
+function PRT!(n::TreeNode, k::Int)
+	
+    if isroot(n)
+        n.data["mcc"] = []
+        for pos in 1:k
+            if !isempty(n.data["child_mccs"][pos]) && length(n.data["child_mccs"][pos])==1
```
Comment (PierreBarrat) [inline src/mcc_branch_labels.jl]: Could this just be `length(n.data["child_mccs"][pos])==1`? If it's empty then it's not of length 1?

File: src/mcc_splits.jl [hunk 4]
```diff
@@ -141,10 +145,25 @@ end
 #=
 Map split `S[i]` to `t`.
 =#
-function _map_split_to_tree(S::SplitList, i::Integer, t::Tree, treesplits::SplitList)
-	# Not lca in case lca(t, leaves(S,i)) contains extra leaves not in S[i]
-	roots = TreeTools.blca([t.lleaves[x] for x in leaves(S,i)]...)
+function _map_split_to_tree(S::SplitList, i::Integer, t::Tree, treesplits::SplitList; strict=false, MCCs=nothing)
```
Comment (PierreBarrat) [inline src/mcc_splits.jl]: This now modifies `t` so we should add  `!` to the name. 
Also, it really only takes `Tree{MiscData}` as input ...

File: src/mcc_splits.jl [hunk 5]
```diff
@@ -81,6 +81,10 @@ end
 function _new_splits!(MCC_splits::SplitList, tree_splits::SplitList)
     idx = Int64[]
     for (i,s) in enumerate(MCC_splits)
+        if isempty(s)
```
Comment (PierreBarrat) [inline src/mcc_splits.jl]: Can this be replaced by 
```
if isempty(s) || in(s, tree_splits, MCC_splits.mask)
   push!(idx,i)
end
```
?

File: src/mcc_branch_labels.jl [hunk 6]
```diff
@@ -0,0 +1,122 @@
+
+"""
+get_mcc_map(MCCs::Vector{Vector{String}}, get_cluster_no =true)
```
Comment (PierreBarrat) [inline src/mcc_branch_labels.jl:3]: docstring does not match function :)

File: src/mcc_branch_labels.jl [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                append!(n.data["mcc"], pop!(n.data["child_mccs"][pos]))
+            else # no unique child MCC and no match with parent -> not part of an MCCs
+                append!(n.data["mcc"], [nothing])
+            end
+        end
+    end
+
+	#delete!(n.data.dat, "child_mccs")
+
+	if !isempty(n.child)
```
Comment (PierreBarrat) [inline src/mcc_branch_labels.jl]: identation issue

**Review by PierreBarrat — COMMENTED** [review]
Tested it with 
```
	t1 = parse_newick_string("((A,B,C,D),E);")
	t2 = parse_newick_string("(((A,B),C),(D,E));")
	MCCs = [["A","B","C"], ["D"], ["E"]]

	tt1 = convert(Tree{MiscData}, t1)
	tt2 = convert(Tree{MiscData}, t2)
	resolve!(tt1, tt2, MCCs; strict=true)
```
which I realized later is the same as the test you added. Seems to work, great! Maybe I want to look at more edge cases later. 

I left some comments in the code, since this is going to be merged with TreeKnit eventually... 
It would be nice if `resolve!` could be called on any type of trees, not just `Tree{MiscData}` as now. It's not urgent to implement this though. 
Also, a "style" comment: I've noticed you use very specific types in function declarations. Sometimes it's just easier not to use any type. But that's a detail.

File: src/mcc_branch_labels.jl (hunk 1 above)
Comment (anna-parker) [reply src/mcc_branch_labels.jl]: Ah yes this is an artifact from Multitreeknit where k was the number of other trees - I will remove!

File: src/mcc_branch_labels.jl (hunk 2 above)
Comment (anna-parker) [reply src/mcc_branch_labels.jl]: Thanks - will change :-)

File: src/mcc_branch_labels.jl (hunk 3 above)
Comment (anna-parker) [reply src/mcc_branch_labels.jl]: Ah yes, I believe so - will also change!

File: src/mcc_branch_labels.jl (hunk 6 above)
Comment (anna-parker) [reply src/mcc_branch_labels.jl:3]: :+1:

File: src/mcc_branch_labels.jl (hunk 7 above)
Comment (anna-parker) [reply src/mcc_branch_labels.jl]: :+1:

File: src/mcc_splits.jl (hunk 5 above)
Comment (anna-parker) [reply src/mcc_splits.jl]: Yes, will change :-)

File: src/mcc_splits.jl (hunk 4 above)
Comment (anna-parker) [reply src/mcc_splits.jl]: Yes that is true - I will change that for now. I think it would be good to discuss if we want to use this form of `assign_mccs` as it alters the tree and needs a tree of `TreeTools.MiscData` not `TreeTools.EmptyData`. If we want to keep it like is, I would suggest either adding a convert tree to MiscData type if `resolve` is called with the `strict=true` argument or altering the default TreeNode type. Otherwise I could rewrite the function next week to use dictionaries and not alter the tree.

**Comment by anna-parker** [comment]
It can occur that the command `TreeKnit.sort_polytomies!` will fail if trees are resolved using strict resolve. Throwing an error, e.g: 

`AssertionError: MCCs 21 and 12 conflict (resp. node internal_117 and internal_67). Are trees resolved using the MCCs?`

This is due to the fact that `sort_polytomies` uses `mcc_map` which cannot have internal nodes that could belong to more than one MCC. Before this is merged into master we need to fix this: e.g. by using `assign_mccs!` instead.

**Comment by anna-parker** [comment]
I have also realized switching to strict resolving in MultiTreeKnit leads my current `fix_consist` functions to fail with the following error:

``` LoadError: KeyError: key nothing not found
Stacktrace:
 [1] getindex(h::Dict{Int64, Vector{String}}, key::Nothing)
   @ Base ./dict.jl:481
 [2] split_MCCs(first::Vector{Vector{String}}, second::Vector{Vector{String}}, third::Vector{Vector{String}}, tree1::Tree{MiscData}, tree2::Tree{MiscData})
``` 
It is unclear to me why this is the case, I feel like they should not be connected unless strict resolve changes the `SplitList`s in an inaccurate way. Before merging I would like to figure out why this fails to make sure there are no hidden problems.

File: src/mcc_splits.jl (hunk 4 above)
Comment (anna-parker) [reply src/mcc_splits.jl]: Separate functions added as of commit: https://github.com/PierreBarrat/TreeKnit.jl/pull/20/commits/dc1a3ddfd69e109578992b53a09769fbf1d0a4d9 .

The current plan is to use `strict_resolve!` for creating nwk files at the end of TreeKnit and still use `resolve!` (with masks) during SA and for creating the ARG. We shall continue to convert trees to `MiscData` objects when `strict_resolve!` is used.

**Comment by anna-parker** [comment]
> It can occur that the command `TreeKnit.sort_polytomies!` will fail if trees are resolved using strict resolve. Throwing an error, e.g:
> 
> `AssertionError: MCCs 21 and 12 conflict (resp. node internal_117 and internal_67). Are trees resolved using the MCCs?`
> 
> This is due to the fact that `sort_polytomies` uses `mcc_map` which cannot have internal nodes that could belong to more than one MCC. Before this is merged into master we need to fix this: e.g. by using `assign_mccs!` instead.

This should be fixed now -> I will try to add a way to convert trees and MCCs to JSON files for viewing in `auspice`.
